🧪 Bump `actions/{upload,download}-artifact` to v4

NilashishC opened the pull request:
##### SUMMARY

They are updated across all GitHub Actions CI/CD workflow definitions
in the repository. It is necessary due to Artifacts API v3 having been
deprecated on Jan 30, 2025.

As a part of the change, the uploaded artifact names needed to be made
unique and their download counterparts had to be invoked against a
name pattern, merging multiple artifacts together.

Making the names unique came with challenges that this patch overcame
by appending SHA-256 hashes of reusable workflow inputs to the
artifact names.

This change also bumps `re-actors/checkout-python-sdist` to v2 as it
uses `actions/download-artifact` internally and so we
needed to get the version that uses non-deprecated APIs.

[1]: https://github.blog/changelog/2024-04-16-deprecation-notice-v3-of-the-artifact-actions/

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Maintenance Pull Request

NilashishC committed:
🧪 Bump actions/{upload,download}-artifact to v4

They are updated across all GitHub Actions CI/CD workflow definitions
in the repository. It is necessary due to Artifacts API v3 having been
deprecated on Jan 30, 2025.

As a part of the change, the uploaded artifact names needed to be made
unique and their download counterparts had to be invoked against a
name pattern, merging multiple artifacts together.

Making the names unique came with challenges that this patch overcame
by appending SHA-256 hashes of reusable workflow inputs to the
artifact names.

[1]: https://github.blog/changelog/2024-04-16-deprecation-notice-v3-of-the-artifact-actions/

Co-Authored-By: Sviatoslav Sydorenko <webknjaz@redhat.com>

Signed-Off-By: NilashishC <nilashishchakraborty8@gmail.com>
Signed-Off-By: Sviatoslav Sydorenko <webknjaz@redhat.com>

NilashishC committed:
📝 Add a change note for PR #676

Co-Authored-By: Sviatoslav Sydorenko <webknjaz@redhat.com>

Signed-Off-By: NilashishC <nilashishchakraborty8@gmail.com>
Signed-Off-By: Sviatoslav Sydorenko <webknjaz@redhat.com>

NilashishC commented:
@webknjaz Hi. Could you please have a look at this PR? This needs to be merged before https://github.com/ansible/pylibssh/pull/675.

Thank you.

webknjaz reviewed (commented):
@NilashishC this is a move in the right direction. It is, however, incomplete.

You have to upgrade `re-actors/checkout-python-sdist@release/v1` to `re-actors/checkout-python-sdist@release/v2` (because v1 uses an older version of the download action).

And additionally, the wheel uploads must change their artifact names to be unique (example: https://github.com/aio-libs/yarl/blob/07c1b4f/.github/workflows/reusable-build-wheel.yml#L92C15-L95C28).

Finally, the download step in the publishing jobs must change to merge+pattern (example: https://github.com/aio-libs/yarl/blob/07c1b4f/.github/workflows/ci-cd.yml#L532C9-L533C29). The last one will not be testable in the PR CI runs and will only fail later if you don't get it right.

NilashishC committed:
🧪 Bump `re-actors/checkout-python-sdist` to v2

This action uses `actions/download-artifact` internally and so we
needed to get the version that uses non-deprecated APIs.

Signed-off-by: NilashishC <nilashishchakraborty8@gmail.com>

NilashishC commented:
@webknjaz thanks for the review.

The variables required to make the wheel name unique in `reusable-build-wheel.yml` seems to be commented out in `inputs` - https://github.com/ansible/pylibssh/blob/devel/.github/workflows/reusable-build-wheel.yml#L17-L26. Should I uncomment those and add it to the wheel name? Or should I choose other inputs.

NilashishC commented:
> @webknjaz thanks for the review.
> 
> The variables required to make the wheel name unique in `reusable-build-wheel.yml` seems to be commented out in `inputs` - https://github.com/ansible/pylibssh/blob/devel/.github/workflows/reusable-build-wheel.yml#L17-L26. Should I uncomment those and add it to the wheel name? Or should I choose other inputs.

Okay, I did try something to make it unique - https://github.com/ansible/pylibssh/pull/676/files#diff-3826dffee70012c1e07ee3d63a4cf5ba7ed82c17bbac68145432a97e71dbade7R187-R192.

However, now some of the build jobs in the matrix fail.

webknjaz commented:
@NilashishC your commit message says

> use merge+pattern for download artifact

But that commit only makes use of the merge input and doesn't replace name with pattern. So these jobs only download the sdist while they expect the wheels to be present.

There might be other problems (I saw some weirdness in the logs) but I would imagine that those are unrelated. Fix this first and we'll see how it goes then.

NilashishC commented:
@webknjaz Apologies for my lack of understanding here. But I'm unsure what the pattern would be . The patterns differ in `reusable-cibuildwheel`  vs `reusable-build-wheel` (because the `inputs` are different). How do you recommend we handle this in the download steps?

webknjaz commented:
Wouldn't just adding an asterisk at the end do it? This should work for as long as the prefix remains the same.

NilashishC commented:
So, all x86 ones are passing, but not the other architectures.

webknjaz commented on .github/workflows/ci-cd.yml:
```diff
@@ -275,7 +275,7 @@ jobs:
     - name: Set up pip cache
       if: >-
         steps.request-check.outputs.release-requested != 'true'
-      uses: actions/cache@v4.0.2
```
Let's move this to a separate PR.

webknjaz commented on .github/workflows/reusable-cibuildwheel.yml:
```diff
@@ -81,9 +81,11 @@ jobs:
-        name: ${{ inputs.dists-artifact-name }}
+        name: ${{ inputs.dists-artifact-name }}-
+          ${{ inputs.os }}-
+          ${{ inputs.qemu }}
```
It might be useful to take this input into account, as #562 revealed the need to run this workflow several times to split wheel building into smaller portions.
```suggestion
          ${{ inputs.qemu }}-
          ${{ inputs.wheel-tags-to-skip }}
```

In the future, though, I'd look into rewriting this to use something like a hash of the entire `${{ inputs }}` object. But that's out of the scope right now.

webknjaz commented:
@NilashishC will you be cleaning up the commits?

webknjaz commented:
> So, all x86 ones are passing, but not the other architectures.

As for https://github.com/ansible/pylibssh/actions/runs/13268100337/job/37040748168?pr=676#step:9:50, it's probably a separate issue that I'd look into separately from this PR.

webknjaz commented on .github/workflows/reusable-tests.yml:
```diff
@@ -190,7 +190,7 @@ jobs:
-      uses: actions/download-artifact@v3
+      uses: actions/download-artifact@v4
       with:
         name: ${{ inputs.dists-artifact-name }}
         path: dist/
```
The tests jobs are failing because this is missing:
```suggestion
        merge-multiple: true
        path: dist/
        pattern: ${{ inputs.dists-artifact-name }}*
```

(I checked the entire PR and this should be the only place needing the change)

NilashishC commented:
@webknjaz I _think_ I have resolved all your comments now. Here's the PR for `actions/cache` - https://github.com/ansible/pylibssh/pull/677.

NilashishC commented:
> @NilashishC will you be cleaning up the commits?

Do you want me to squash them into a single one?

webknjaz commented:
> Do you want me to squash them into a single one?

It's up to you now. I like a series of atomic commits. Not a blind squash. Currently, the repo still allows the fake merge methods like “squash” or “rebase” but I'm eventually going to enable the merge queues feature which will take those away.
So if you want this to be accepted via the squash method, it's fine for now. Just make sure there's nothing unrelated to the title of the PR in the diff.

webknjaz commented on .github/workflows/reusable-build-wheel.yml:
```diff
@@ -184,9 +184,12 @@ jobs:
-        name: ${{ inputs.dists-artifact-name }}
+        name: ${{ inputs.dists-artifact-name }}-
+          ${{ inputs.manylinux-year-target }}-
+          ${{ inputs.manylinux-image-target-arch }}-
```
Looks like one more factor is missing here:
```suggestion
          ${{ inputs.manylinux-image-target-arch }}-
          ${{ inputs.manylinux-image-target-qemu-arch }}-
```

It's not breaking the CI in PR builds because a portion of jobs are being skipped. However, it'll likely explode on release.

webknjaz commented:
Urgh.. https://github.com/ansible/pylibssh/actions/runs/13306998592/job/37160303882?pr=676#step:6:15 fails due to asterisks. Looks like that inputs hashing idea will probably have to make its way into this PR's scope. I'll think about it and maybe make a direct commit into your branch.

webknjaz commented:
@NilashishC I merged this, but the CI still needs love to be able to function and make releases. I'm planning to keep looking into it. Stay tuned!
